pkg/logs/tekton_logging.go

```diff
@@ -40,12 +41,15 @@ type TektonLogger struct {
 	LogWriter         LogWriter
 	Namespace         string
 	LogsRetrieverFunc retrieverFunc
+	logsChannel       chan LogLine
+	errorsChannel     chan error
+	wg                *sync.WaitGroup
 }
 
 // LogWriter is an interface that can be implemented to define different ways to stream / write logs
 // it's the implementer's responsibility to route those logs through the corresponding medium
 type LogWriter interface {
-	WriteLog(line LogLine) error
+	WriteLog(line LogLine, lch chan<- LogLine) error
```

Comment:
I find this interface very odd. What is this function really saying? Hey LogWriter, I want you do log a line and btw, here is the channel to write it to? Would you in this case not just send the log message to the channel yourself and be done with it? Looking at the implementations of this interface this is pretty much what is happening now.

I still argue if one really wants to decouple the actual log writing from what else is happening in tekton logger, it should be the responsibility of the LogWriter to provide a channel and tekton_logging just sends everything down that sink. This way you don't have to deal with all the goroutines and synchronisation in tekton_logigng.

Or maybe alternatively you have a single Stream function which takes a channel and reads from it until it closes. In this case, tekton_logging would create the channel and call Stream, but then just this channel, no differentiation between WriteLog and StreamLog. 

Either way, if you want to use channel there should only be a single function in the interface. One needs to decide who would be responsible for the channel creation LogWriter or TektonLogger. Either way, all logging would need to go through the channel.

Reply:
Talking about removing `WriteLog`. I think it would make sense to not print the header line for the container at all:

```
err = t.LogWriter.WriteLog(LogLine{
			Line: fmt.Sprintf("\nShowing logs for build %v stage %s and container %s\n",				Line: fmt.Sprintf("\nShowing logs for build %v stage %s and container %s",
				infoColor.Sprintf(buildName), infoColor.Sprintf(stageName), infoColor.Sprintf(ic.Name)),					infoColor.Sprintf(buildName), infoColor.Sprintf(stageName), infoColor.Sprintf(ic.Name)),
```

I am thinking about https://github.com/jenkins-x/jx/issues/5169 in particular as well and I think it makes sense for the UI as well. 

Instead, we could enhance `LogLine` with contextual information about the current pod, container, stage, etc. Then it becomes a concern of the `LogWriter` to decide how it wants to render the logs from the various containers and how it wants to separate the logs. The `LogWriter` can decide to write a header logline as we do now, but it could also prepend contextual information to each log line it prints. One would get a solution for #5169 pretty much for free by just plugging in a different implementation of `LogWriter`. 

We can add to the contextual information whether a given logline is from a failed step/stage. So again it becomes the `LogWriter` who is responsible for adding colour escape sequences. When refactoring the tests and matching expected log lines, I, for example, had to account escape sequences which was odd. 

IMO we would end up with a more flexible solution, a `LogWriter` which actually does have responsibilities as well as better separation of concerns. @dgozalo, @abayer WDYT?

Reply:
Hmm, that seems reasonable to me. One thing I know I'd like eventually is timestamping on the log lines, so contextual info may be the only way to get there.

Reply:
Yeah, I eventually wanted to surfase the tekton logging format and see if it would fit us better.

The idea of sending a `LogLine` struct as an event was exactly that, we'd probably want to add some metadata to the traces at some point, and that was just the way to set it up.

I think we should try to get the opinion of the community on the formatting, though.

Reply:
Also, I kind of agree on the interface being a bit weird right now, and it all comes down to the fixes that were needed due to synchronization.

I think it would make sense to just keep the StreamLog function. You can get every single line sent to the channel there too, in case you want to add some extra metadata.

I think the creation of the channels and synchronization should be in tekton_logging, that way you make the Writers a lot of simple to plug and use.

Reply:
> Also, I kind of agree on the interface being a bit weird right now, and it all comes down to the fixes that were needed due to synchronization.

Cool

> I think it would make sense to just keep the StreamLog function. You can get every single line sent to the channel there too, in case you want to add some extra metadata.

+1

> I think the creation of the channels and synchronization should be in tekton_logging, that way you make the Writers a lot simpler to plug and use.

I guess that might be a matter of taste. The points above are more important to agree on.